File: code_blocks/tools/paywalls_on_purchase_initiated_1.swift
PaywallView()
    .onPurchaseInitiated({ package, resume in
        print("User is attempting to purchase \(package)")
        Task {
            do {
                try await performAuthentication()
                print("Authentication Complete, presenting purchase flow")
                resume()
            } catch {
                print("Authentication failed… cancel purchase flow")
                resume(shouldProceed: false)
                presentErrorAlertForAuthentication()
            }
        }
    })
    // in iOS we can also intercept offer code redemption flows
    .onOfferCodeRedemptionInitiated({ resume in
        print("Offer code redemption initiated")
        Task {
            do {
                try await performAuthentication()
                print("Authentication Complete, presenting redeem flow")
                resume()
            } catch {
                print("Authentication failed… cancel redeem flow")
                resume(shouldProceed: false)
                presentErrorAlertForAuthentication()
            }
        }
    })

File: code_blocks/tools/paywalls_on_purchase_initiated_1.kt
@OptIn(ExperimentalPreviewRevenueCatUIPurchasesAPI::class)
@Composable
private fun LockedScreen() {
    YourContent()

    PaywallDialog(
        PaywallDialogOptions.Builder()
            .setRequiredEntitlementIdentifier(Constants.ENTITLEMENT_ID)
            .setShouldDisplayDismissButton(false)
            .setListener(
                object : PaywallListener {
                    override fun onPurchasePackageInitiated(rcPackage: Package, resume: Resumable) { 
                        println("on purchase initiated for $rcPackage")
                        performAuthentication { result -> 
                            when (result) {
                                is Success -> resume().also { 
                                    println("Authentication complete, presenting purchase flow")
                                }
                                is Failure -> resume(false).also {
                                    println("Authentication failed, canceling purchase flow")
                                    presentErrorAlert()
                                }
                            }
                        }
                    }
                }
            )
            .build()
    )
}
